velox/functions/prestosql/ArrayShuffle.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <numeric>
+#include <random>
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+// See documentation at
+// https://prestodb.io/docs/current/functions/array.html#shuffle
+class ArrayShuffleFunction : public exec::VectorFunction {
+ public:
+  bool isDeterministic() const override {
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /*outputType*/,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    VectorPtr localResult;
+
+    // Input can be constant, flat or dictionary.
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatBaseArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
+      singleRow->clearAll();
+      singleRow->setValid(flatIndex, true);
+      singleRow->updateBounds();
+
+      localResult = shuffleDecodedElementIndices(
+          *singleRow.get(), flatBaseArray, context);
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      // For deterministic single-arg functions, expression evaluation will peel
+      // off encodings and guarantee that we will only see flat or constant
+      // inputs. But this guarantee does not exist for non-deterministic
+      // functions like shuffle(). Hence, we need to use DecodedVector to handle
+      // the possible dict encoding.
+      localResult = shuffleDecodedElementIndices(rows, arg, context);
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  VectorPtr shuffleDecodedElementIndices(
+      const SelectivityVector& rows,
+      const VectorPtr& inputVector,
+      exec::EvalCtx& context) const {
+    exec::LocalDecodedVector localDecodedVector(context, *inputVector, rows);
+    const auto& decodedArray = *localDecodedVector.get();
+    auto flatBaseArray = decodedArray.base()->as<ArrayVector>();
```

Comment:
nit: consider renaming flatBaseArray and baseArrayElements to arrayVector and elementsVector for clarity.